src/cargo/sources/registry/index.rs

```diff
@@ -698,10 +700,13 @@ impl<'cfg> RegistryIndex<'cfg> {
         pkg: PackageId,
         load: &mut dyn RegistryData,
     ) -> Poll<CargoResult<bool>> {
-        let req = OptVersionReq::exact(pkg.version());
-        let found = ready!(self.summaries(pkg.name(), &req, load))?
-            .filter(|s| s.package_id().version() == pkg.version())
-            .any(|s| s.is_yanked());
+        let mut req = OptVersionReq::exact(pkg.version());
+
+        // Since crates.io allows crate versions to differ only by build metadata,
+        // a query using OptVersionReq::exact + next() can return nondeterministic results.
+        // So we `lock_to` the exact version were interested in
+        req.lock_to(pkg.version());
```

Comment:
nit: `exact` + `lock_to` seems like a weird dance.  Should we have a new constructor for this purpose?

Reply:
Yes. Sounds good.